src/ripple/protocol/impl/STTx.cpp

```diff
@@ -236,15 +237,31 @@ Json::Value STTx::getJson(JsonOptions) const
 Json::Value
 STTx::getJson(JsonOptions options, bool binary) const
 {
+    if (!(options & JsonOptions::disable_API_prior_V2))
+    {
+        // Behaviour before API version 2
+        if (binary)
+        {
+            Json::Value ret;
+            Serializer s = STObject::getSerializer();
+            ret[jss::tx] = strHex(s.peekData());
+            ret[jss::hash] = to_string(getTransactionID());
+            return ret;
+        }
+        return getJson(options);
+    }
+
+    // API version 2 behaviour
     if (binary)
     {
-        Json::Value ret;
         Serializer s = STObject::getSerializer();
-        ret[jss::tx] = strHex(s.peekData());
-        ret[jss::hash] = to_string(getTransactionID());
+        Json::Value ret = strHex(s.peekData());
         return ret;
     }
-    return getJson(options);
+    // We want virtualy the same output as `getJson(JsonOptions)` overload
+    // above, but without the hash element. Since `getJson(JsonOptions)`
+    // is using STObject::getJson(JsonOptions::none), we use it here as well
+    return STObject::getJson(JsonOptions::none);
```

Comment:
The proposed code is incorrect - in API version 2 we are not storing binary inside `tx` element, instead we return it directly (just like we return non-binary). As for returning transaction id - the way I wrote it, the decision to use transaction id (and/or any other field we need, and indeed we do need ledger index) belongs to the caller. No need to expand the return value; just use the accessors which are right there. And yes I do not like accessors either, but fixing _that_ belongs to a refactoring PR rather than this one.

However, I do see what you mean by _almost_ duplicating the two options, that indeed can be confusing to the reader.

Reply:
ok rewrote it to reduce code duplication, the only duplicate part now is setting `ret[jss::hash] = to_string(getTransactionID());` to support API version 1. Hopefully one day we can rid of it.